src/docker.c

```diff
@@ -0,0 +1,1137 @@
+      old_stats->old_percpu_usage[i] = stats->percpu_usage[i];
+      old_stats->old_system_usage = stats->system_cpu_usage;
+    }
+  } else {
+    cpu_state_t *old_stats = (cpu_state_t *) calloc(1, sizeof(cpu_state_t));
+    old_stats->num_cpus = stats->num_cpus;
+    old_stats->old_system_usage = stats->system_cpu_usage;
+    old_stats->old_percpu_usage =
+        (unsigned long *) calloc(stats->num_cpus, sizeof(unsigned long));
+    for (int i = 0; i < stats->num_cpus; i++) {
+      old_stats->old_percpu_usage[i] = stats->percpu_usage[i];
+      stats->percpu_percent_used[i] = stats->system_cpu_usage > 0 ?
+          (100.00*stats->percpu_usage[i])/(1.00*stats->system_cpu_usage) : 0.00;
+      stats->percpu_percent_idle[i] = 100.00 - stats->percpu_percent_used[i];
+      stats->percpu_idle[i] = stats->system_cpu_usage - stats->percpu_usage[i];
+    }
+    if (c_avl_insert(cpu_hist_values, (void *) container_id,
+           (void *) old_stats) < 0) {
+      ERROR("docker: c_avl_insert failed!");
+    }
+  }
+}
+
+// Computes percentages for Memory metrics in different states.
+static void compute_memory_stats(memory_stats_t *stats) {
+  stats->free = stats->limit - stats->usage;
+  stats->used_percentage = (stats->usage * 100.00)/(stats->limit);
+  stats->free_percentage = 100.00 - stats->used_percentage;
+}
+
+//==============================================================================
+//==============================================================================
+//==============================================================================
+// Functionality to parse the Docker Stats JSON response and retrieve stats.
+//==============================================================================
+//==============================================================================
+//==============================================================================
+
+// Retrieves the list of 64 character container IDs from the containers/json
+// endpoint.
+static int extract_container_ids_from_response(char ***container_list,
+    char *response_buffer) {
+  yajl_val node;
+  char errbuf[1024];
+  int num_containers = -1;
+  char **list;
+  node = yajl_tree_parse(response_buffer, errbuf, sizeof(errbuf));
+  if (node == NULL) {
+    if (strlen(errbuf)) {
+      ERROR("docker: parse_error: %s.\n", errbuf);
+    } else {
+      ERROR("docker: extract_container_ids_from_response: parse_error.\n");
+    }
+    goto error;
+  }
+  if (node->u.array.len == 0) {
+    return 0;
+  }
+
+  const char *id_path[] = { "Id", (const char *) 0 };
+  if (YAJL_IS_ARRAY(node)) {
+    list = (char **) calloc(node->u.array.len, sizeof(char *));
+    if (list == NULL) {
+      ERROR("docker: extract_container_ids_from_response: malloc failed!");
+    }
+    num_containers = node->u.array.len;
+    for (int i = 0; i < num_containers; i++) {
+      yajl_val elem = node->u.array.values[i];
```

Comment:
Do we need yajl_tree_free(elem) and yajl_tree_free(id_node) here?

If yes, please check other places where yajl_val variables are declared to make sure they are freed as well.